File: minigit-4/ts/src/commands/add.ts
// add command - Stage files for commit

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot, normalizePathSeparator, getFileModeFromStat } from '../utils';
import { readIndex, writeIndex, createIndexEntryFromFile, addToIndex, removeFromIndex } from '../index-file';
import { createBlobFromFile, createBlobFromSymlink } from '../objects';

export function add(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  let all = false;
  let update = false;
  const paths: string[] = [];

  for (const arg of args) {
    if (arg === '-A' || arg === '--all') {
      all = true;
    } else if (arg === '-u' || arg === '--update') {
      update = true;
    } else {
      paths.push(arg);
    }
  }

  let entries = readIndex(repoRoot);

  if (all) {
    // Stage all changes (new, modified, deleted)
    entries = addAllChanges(repoRoot, entries);
  } else if (update) {
    // Stage only tracked file modifications and deletions
    entries = updateTrackedFiles(repoRoot, entries);
  } else if (paths.length === 0) {
    console.error('fatal: no pathspec given');
    return 1;
  } else {
    // Stage specific paths
    for (const p of paths) {
      const result = addPath(repoRoot, entries, p);
      if (result.error) {
        console.error(result.error);
        return 1;
      }
      entries = result.entries;
    }
  }

  writeIndex(repoRoot, entries);
  return 0;
}

function addPath(
  repoRoot: string,
  entries: ReturnType<typeof readIndex>,
  pathSpec: string
): { entries: ReturnType<typeof readIndex>; error?: string } {
  const fullPath = path.resolve(repoRoot, pathSpec);

  // Check if path exists
  if (!fs.existsSync(fullPath)) {
    // Check if it was a tracked file that was deleted
    const relativePath = normalizePathSeparator(path.relative(repoRoot, fullPath));
    const existingEntry = entries.find((e) => e.name === relativePath);
    if (existingEntry) {
      // Mark as deleted by removing from index
      entries = removeFromIndex(entries, relativePath);
      return { entries };
    }
    return { entries, error: `fatal: pathspec '${pathSpec}' did not match any files` };
  }

  const stat = fs.lstatSync(fullPath);

  if (stat.isDirectory()) {
    // Add all files in directory recursively
    return addDirectory(repoRoot, entries, fullPath);
  } else if (stat.isSymbolicLink()) {
    // Add symlink
    const relativePath = normalizePathSeparator(path.relative(repoRoot, fullPath));
    const sha = createBlobFromSymlink(repoRoot, fullPath);
    const entry = createIndexEntryFromFile(relativePath, sha, 0o120000, stat);
    entries = addToIndex(entries, entry);
    return { entries };
  } else {
    // Add regular file
    const relativePath = normalizePathSeparator(path.relative(repoRoot, fullPath));
    const sha = createBlobFromFile(repoRoot, fullPath);
    const mode = getFileModeFromStat(stat);
    const entry = createIndexEntryFromFile(relativePath, sha, mode, stat);
    entries = addToIndex(entries, entry);
    return { entries };
  }
}

function addDirectory(
  repoRoot: string,
  entries: ReturnType<typeof readIndex>,
  dirPath: string
): { entries: ReturnType<typeof readIndex>; error?: string } {
  const dirEntries = fs.readdirSync(dirPath, { withFileTypes: true });

  for (const entry of dirEntries) {
    if (entry.name === '.minigit') continue;

    const fullPath = path.join(dirPath, entry.name);

    if (entry.isDirectory()) {
      const result = addDirectory(repoRoot, entries, fullPath);
      if (result.error) return result;
      entries = result.entries;
    } else if (entry.isSymbolicLink()) {
      const stat = fs.lstatSync(fullPath);
      const relativePath = normalizePathSeparator(path.relative(repoRoot, fullPath));
      const sha = createBlobFromSymlink(repoRoot, fullPath);
      const indexEntry = createIndexEntryFromFile(relativePath, sha, 0o120000, stat);
      entries = addToIndex(entries, indexEntry);
    } else if (entry.isFile()) {
      const stat = fs.statSync(fullPath);
      const relativePath = normalizePathSeparator(path.relative(repoRoot, fullPath));
      const sha = createBlobFromFile(repoRoot, fullPath);
      const mode = getFileModeFromStat(stat);
      const indexEntry = createIndexEntryFromFile(relativePath, sha, mode, stat);
      entries = addToIndex(entries, indexEntry);
    }
  }

  return { entries };
}

function addAllChanges(repoRoot: string, entries: ReturnType<typeof readIndex>): ReturnType<typeof readIndex> {
  // Get all tracked files from index
  const trackedPaths = new Set(entries.map((e) => e.name));

  // Remove deleted files from index
  entries = entries.filter((e) => {
    const fullPath = path.join(repoRoot, e.name);
    return fs.existsSync(fullPath);
  });

  // Add/update all files in working directory
  const result = addDirectory(repoRoot, entries, repoRoot);
  return result.entries;
}

function updateTrackedFiles(repoRoot: string, entries: ReturnType<typeof readIndex>): ReturnType<typeof readIndex> {
  const newEntries: typeof entries = [];

  for (const entry of entries) {
    const fullPath = path.join(repoRoot, entry.name);

    if (!fs.existsSync(fullPath)) {
      // File deleted - don't include in new entries
      continue;
    }

    const stat = fs.lstatSync(fullPath);

    if (stat.isSymbolicLink()) {
      const sha = createBlobFromSymlink(repoRoot, fullPath);
      const newEntry = createIndexEntryFromFile(entry.name, sha, 0o120000, stat);
      newEntries.push(newEntry);
    } else {
      const sha = createBlobFromFile(repoRoot, fullPath);
      const mode = getFileModeFromStat(stat);
      const newEntry = createIndexEntryFromFile(entry.name, sha, mode, stat);
      newEntries.push(newEntry);
    }
  }

  return newEntries;
}

File: minigit-4/py/commands/add.py
"""add command - Stage files for commit."""

import sys
import os
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root, get_file_mode
from objects import hash_object
from index import Index, IndexEntry


def run(args):
    """Stage files for commit."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    if not args:
        print("Nothing specified, nothing added.", file=sys.stderr)
        return 0

    # Parse flags
    all_flag = False
    update_flag = False
    paths = []

    i = 0
    while i < len(args):
        arg = args[i]
        if arg in ('-A', '--all'):
            all_flag = True
        elif arg in ('-u', '--update'):
            update_flag = True
        elif arg == '.':
            paths.append(repo_root)
        else:
            paths.append(Path.cwd() / arg)
        i += 1

    # Read existing index
    index = Index.read(repo_root)

    # Handle -A flag (stage all changes including deletions)
    if all_flag:
        paths = [repo_root]
        update_flag = False  # -A overrides -u

    # Handle -u flag (only update tracked files)
    if update_flag:
        # Stage modifications and deletions of tracked files
        tracked_files = set(index.entries.keys())
        for name in tracked_files:
            file_path = repo_root / name
            if file_path.exists():
                # File still exists - check if modified
                try:
                    content = file_path.read_bytes()
                    sha = hash_object(content, 'blob', write=True, repo_root=repo_root)
                    entry = IndexEntry.from_file(file_path, sha, repo_root)
                    index.add_entry(entry)
                except Exception as e:
                    print(f"error: {name}: {e}", file=sys.stderr)
            else:
                # File deleted
                index.remove_entry(name)
        index.write(repo_root)
        return 0

    # Collect all files to add
    files_to_add = []
    for path in paths:
        path = Path(path).resolve()

        if not path.exists():
            # Check if it's a tracked file that was deleted
            try:
                rel_path = str(path.relative_to(repo_root))
                if rel_path in index.entries:
                    if all_flag:
                        index.remove_entry(rel_path)
                        continue
                    else:
                        print(f"fatal: pathspec '{args[0]}' did not match any files", file=sys.stderr)
                        return 1
            except ValueError:
                pass
            print(f"fatal: pathspec '{path}' did not match any files", file=sys.stderr)
            return 1

        if path.is_file():
            files_to_add.append(path)
        elif path.is_dir():
            # Recursively add all files in directory
            for file_path in path.rglob('*'):
                if file_path.is_file():
                    # Skip .minigit directory
                    try:
                        rel = file_path.relative_to(repo_root)
                        if '.minigit' in rel.parts:
                            continue
                    except ValueError:
                        continue
                    files_to_add.append(file_path)

            # Also handle deletions if -A flag
            if all_flag:
                existing_files = set()
                for f in path.rglob('*'):
                    if f.is_file():
                        try:
                            rel = str(f.relative_to(repo_root))
                            if '.minigit' not in rel:
                                existing_files.add(rel)
                        except ValueError:
                            pass

                # Find deleted files
                for name in list(index.entries.keys()):
                    file_path = repo_root / name
                    if not file_path.exists():
                        index.remove_entry(name)

    # Add each file to index
    for file_path in files_to_add:
        try:
            # Read file content
            if file_path.is_symlink():
                content = os.readlink(file_path).encode()
            else:
                content = file_path.read_bytes()

            # Hash and store the blob
            sha = hash_object(content, 'blob', write=True, repo_root=repo_root)

            # Create index entry
            entry = IndexEntry.from_file(file_path, sha, repo_root)
            index.add_entry(entry)

        except Exception as e:
            print(f"error: {file_path}: {e}", file=sys.stderr)
            return 1

    # Handle deletions for -A
    if all_flag:
        for name in list(index.entries.keys()):
            file_path = repo_root / name
            if not file_path.exists():
                index.remove_entry(name)

    # Write updated index
    index.write(repo_root)
    return 0
